csrc/multidevice/communication.cpp

```diff
@@ -111,209 +98,225 @@ Broadcast::Broadcast(CommParams params) : Communication(params, "broadcast") {}
+      ->broadcast(tensors, {.rootRank = root_relative_index_});
 }
 
-Gather::Gather(CommParams params) : Communication(params, "gather") {
-  assertBufferCount(params_.src_bufs, 1);
-}
+Gather::Gather(CommParams params) : Communication(params, "gather") {}
 
 c10::intrusive_ptr<c10d::Work> Gather::post(
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
   post_common(*this, comm);
-  // This is used to change the representation of the buffers to match c10d
-  // ProcessGroup API
-  std::vector<std::vector<at::Tensor>> buf_list;
-  if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.dst_bufs, params_.team.size());
-    buf_list = {std::move(params_.dst_bufs)};
-  } else {
-    assertBufferCount(params_.dst_bufs, 0);
+
+  if (comm.deviceId() == params_.root && !params_.is_root_in_mesh) {
+    input_tensor = at::empty_like(output_tensor.slice(0, 0, 1));
   }
-  auto work =
-      comm.getBackendForTeam(params_.team, backend)
-          ->gather(
-              buf_list, params_.src_bufs, {.rootRank = root_relative_index_});
+  std::vector<at::Tensor> input_tensors({input_tensor});
+
+  std::vector<std::vector<at::Tensor>> output_tensors;
   if (comm.deviceId() == params_.root) {
-    params_.dst_bufs = std::move(buf_list.back());
+    output_tensors.resize(1);
+    int64_t j = 0;
+    for (auto i : c10::irange(params_.team.size())) {
+      if (root_relative_index_ == static_cast<DeviceIdxType>(i) &&
+          !params_.is_root_in_mesh) {
+        output_tensors.front().push_back(input_tensor);
+        continue;
+      }
+      output_tensors.front().push_back(output_tensor.slice(0, j, j + 1));
+      j++;
+    }
   }
-  return work;
+
+  return comm.getBackendForTeam(params_.team, backend)
+      ->gather(
+          output_tensors, input_tensors, {.rootRank = root_relative_index_});
 }
 
 Allgather::Allgather(CommParams params)
-    : Communication(params, "allgather", false) {
-  assertBufferCount(params_.src_bufs, 1);
-  assertBufferCount(params_.dst_bufs, params_.team.size());
-}
+    : Communication(params, "allgather", false) {}
 
 c10::intrusive_ptr<c10d::Work> Allgather::post(
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
   post_common(*this, comm);
-  // This is used to change the representation of the buffers to match c10d
-  // ProcessGroup API
-  std::vector<std::vector<at::Tensor>> buf_list;
-  buf_list = {std::move(params_.dst_bufs)};
-  auto work = comm.getBackendForTeam(params_.team, backend)
-                  ->allgather(buf_list, params_.src_bufs, {});
-  params_.dst_bufs = std::move(buf_list.back());
-  return work;
-}
 
-Scatter::Scatter(CommParams params) : Communication(params, "scatter") {
-  assertBufferCount(params_.dst_bufs, 1);
+  std::vector<at::Tensor> input_tensors({input_tensor});
+
+  // FIXME: use split
```

Comment:
what do you mean by "use split"? (for my own understanding)

Reply:
`at::split` instead of a for-loop of `at::slice`. But I'll be aware of your other comment on device order in mesh (I haven't got a chance to look further yet).